src/platform/i386/ioapic.c

```diff
@@ -0,0 +1,195 @@
+{
+        IOAPIC_POL_ACTHIGH = 0,
+        IOAPIC_POL_ACTLOW  = 1,
+};
+
+enum ioapic_triggermode
+{
+        IOAPIC_TRIGGER_EDGE  = 0,
+        IOAPIC_TRIGGER_LEVEL = 1,
+};
+
+struct ioapic_info {
+	unsigned int   ioapicid;
+	volatile void *io_vaddr;
+	int            nentries;
+	int            glbint_base;
+};
+
+static struct ioapic_info ioapicinfo[IOAPIC_MAX] = { { 0, NULL, 0, 0} };
+static unsigned int ioapic_count;
+
+void
+ioapic_set_page(struct ioapic_info *io, u32_t page)
+{
+        io->io_vaddr = (volatile u32_t *)(page * (1 << 22) | ((u32_t)io->io_vaddr & ((1 << 22) - 1)));
+
+        printk("\tSet IOAPIC %d @ %p\n", io->ioapicid, io->io_vaddr);
+}
+
+static void
+ioapic_reg_write(struct ioapic_info *io, u8_t offset, u32_t val)
+{
+        *(volatile u32_t *)(io->io_vaddr + IOAPIC_IOREGSEL) = offset;
+        *(volatile u32_t *)(io->io_vaddr + IOAPIC_IOWIN)    = val;
+}
+
+static u32_t
+ioapic_reg_read(struct ioapic_info *io, u8_t offset)
+{
+        *(volatile u32_t *)(io->io_vaddr + IOAPIC_IOREGSEL) = offset;
+
+        return *(volatile u32_t *)(io->io_vaddr + IOAPIC_IOWIN);
+}
+
+void
+ioapic_int_mask(int intnum)
+{
+	/* TODO */
+}
+
+void
+ioapic_int_unmask(int intnum)
+{
+	/* TODO */
+}
+
+static struct ioapic_info *
+ioapic_findbygsi(int irq)
+{
+	unsigned int i = 0;
+
+	for (; i < ioapic_count; i++) {
+		if (irq >= ioapicinfo[i].glbint_base && irq < ioapicinfo[i].nentries) return &ioapicinfo[i];
+	}
+
+	return NULL;
+}
+
+static struct ioapic_info *
+ioapic_findbyid(int id)
+{
+	unsigned int i = 0;
+
+	for (; i < ioapic_count; i++) {
+		if (id == (int)(ioapicinfo[i].ioapicid)) return &ioapicinfo[i];
+	}
+
+	return NULL;
+}
+
+void
+ioapic_int_override(struct intsrcovrride_cntl *iso)
+{
+	assert(iso->header.len == sizeof(struct intsrcovrride_cntl));
+
+	if (iso->source != iso->glb_int_num_off) {
+		struct ioapic_info *ioap = ioapic_findbygsi(iso->glb_int_num_off);
+
+		assert(ioap);
+		printk("\tInterrupt Source Override for [%u] => %u with IOAPIC %d\n", iso->source, iso->glb_int_num_off, ioap->ioapicid);
+		ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(iso->glb_int_num_off), iso->source + 32);
+	}
+}
+
+void
+ioapic_int_enable(int irqnum, int cpunum, int addflag)
+{
+	struct ioapic_info *ioap = ioapic_findbygsi(irqnum);
+
+	assert(ioap);
+	if (addflag) {
+		/* TODO: logical destination = 1 and add core no or lapic number? */
+	} else {
+		ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum), irqnum + 32);
+		ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum)+1, cpunum<<24);
+	}
+}
+
+void
+ioapic_int_disable(int irqnum)
+{
+	struct ioapic_info *ioap = ioapic_findbygsi(irqnum);
+
+	assert(ioap);
+        ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum), IOAPIC_INT_DISABLED | irqnum);
+        ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum)+1, 0);
+}
+
+void
+ioapic_iter(struct ioapic_cntl *io)
+{
+	u32_t ver;
+	int ioent, j;
+	static int more = 0;
+
+	assert(io);
+
+	if (ioapic_count == IOAPIC_MAX) {
```

Comment:
check should be `>=` for that log to print for every item over MAX. :P will fix.